atomix/cluster/src/test/java/io/atomix/raft/RaftRule.java

```diff
@@ -0,0 +1,521 @@
+  private final AtomicReference<CommitAwaiter> commitAwaiterRef = new AtomicReference<>();
+  private long position;
+
+  private RaftRule(final int nodeCount) {
+    this.nodeCount = nodeCount;
+  }
+
+  public static RaftRule withBootstrappedNodes(final int nodeCount) {
+    if (nodeCount < 1) {
+      throw new IllegalArgumentException("Expected to have at least one node to configure.");
+    }
+    return new RaftRule(nodeCount);
+  }
+
+  public static RaftRule withoutNodes() {
+    return new RaftRule(-1);
+  }
+
+  @Override
+  public Statement apply(final Statement base, final Description description) {
+    final var statement = super.apply(base, description);
+    return temporaryFolder.apply(statement, description);
+  }
+
+  @Override
+  protected void before() throws Throwable {
+    directory = temporaryFolder.newFolder().toPath();
+
+    position = 0;
+    members = new ArrayList<>();
+    memberLog = new ConcurrentHashMap<>();
+    nextId = 0;
+    context = new SingleThreadContext("raft-test-messaging-%d");
+    protocolFactory = new TestRaftProtocolFactory(context);
+
+    if (nodeCount > 0) {
+      createServers(nodeCount);
+    }
+  }
+
+  @Override
+  protected void after() {
+    servers.forEach(
+        s -> {
+          try {
+            if (s.isRunning()) {
+              s.shutdown().get(30, TimeUnit.SECONDS);
+            }
+          } catch (final Exception e) {
+            // its fine..
+          }
+        });
+    servers.clear();
+    context.close();
+    context = null;
+    members.clear();
+    nextId = 0;
+    protocolFactory = null;
+    highestCommit = 0;
+    commitAwaiterRef.set(null);
+    memberLog.clear();
+    memberLog = null;
+    position = 0;
+  }
+
+  /**
+   * Returns the next server address.
+   *
+   * @param type The startup member type.
+   * @return The next server address.
+   */
+  private RaftMember nextMember(final RaftMember.Type type) {
+    return new TestMember(nextNodeId(), type);
+  }
+
+  /**
+   * Returns the next unique member identifier.
+   *
+   * @return The next unique member identifier.
+   */
+  private MemberId nextNodeId() {
+    return MemberId.from(String.valueOf(++nextId));
+  }
+
+  /** Creates a set of Raft servers. */
+  private List<RaftServer> createServers(final int nodes) throws Exception {
+    final List<RaftServer> servers = new ArrayList<>();
+
+    for (int i = 0; i < nodes; i++) {
+      members.add(nextMember(RaftMember.Type.ACTIVE));
+    }
+
+    final CountDownLatch latch = new CountDownLatch(nodes);
+
+    for (int i = 0; i < nodes; i++) {
+      final RaftServer server = createServer(members.get(i).memberId());
+      if (members.get(i).getType() == RaftMember.Type.ACTIVE) {
+        server
+            .bootstrap(members.stream().map(RaftMember::memberId).collect(Collectors.toList()))
+            .thenAccept(this::addCommitListener)
+            .thenRun(latch::countDown);
+      } else {
+        server
```

Comment:
When can this happen? Lines 166 look they always create active members

Reply:
this was copied from the previous RaftTests. It is needed when we test with passive nodes. But we can also remove it for now